nodes/stretch_ik_control.py

```diff
@@ -0,0 +1,1217 @@
+            callback_group=MutuallyExclusiveCallbackGroup(),
+        )
+
+        # Get the joint limits
+        ok = self.__load_joint_limits(speed_profile=self.speed_profile)
+        if not ok:
+            return False
+
+        # Create a service client to switch to navigation mode
+        navigation_mode = ControlMode.NAVIGATION
+        self.switch_to_navigation_client = self.node.create_client(
+            Trigger,
+            navigation_mode.get_service_name(),
+            callback_group=MutuallyExclusiveCallbackGroup(),
+        )
+
+        # Create a publisher and action client to command robot motion
+        self.base_vel_pub = self.node.create_publisher(
+            Twist,
+            "/stretch/cmd_vel",
+            qos_profile=1,
+        )
+        self.arm_client_future = None
+        self.arm_client = ActionClient(
+            self.node,
+            FollowJointTrajectory,
+            "/stretch_controller/follow_joint_trajectory",
+            callback_group=MutuallyExclusiveCallbackGroup(),
+        )
+
+        self.initialized = True
+        return True
+
+    def __load_joint_limits(self, speed_profile: SpeedProfile) -> bool:
+        """
+        Loads the position and velocity limits for the robot's joints.
+
+        Note that we load a single limit for the arm, as opposed to separate limits
+        per telescoping joint. Also note that this function blocks until the service
+        to get joint limits is available.
+
+        Parameters
+        ----------
+        speed_profile: The speed profile to use to get max velocities and accelerations.
+
+        Returns
+        -------
+        bool: True if the joint limits were successfully loaded.
+        """
+        # Invoke the service to get joint limits. Note that this service
+        # does not actually return the joint limits. Rather, it publishes the joint
+        # limits to the /joint_limits topic.
+        get_joint_limits_client = self.node.create_client(
+            Trigger,
+            "get_joint_states",
+            # callback_group=MutuallyExclusiveCallbackGroup(),
+        )
+        self.node.get_logger().info("Waiting for the get joint states service...")
+        get_joint_limits_client.wait_for_service()  # TODO: This can block!
+        response = get_joint_limits_client.call(Trigger.Request())
+        if not response.success:
+            self.node.get_logger().error("Failed to get joint limits.")
+            return False
+        self.node.get_logger().info("Received the joint limits.")
+
+        # Get the robot parameters
+        robot_params = RobotParams().get_params()[1]
+        speed_profile_str = speed_profile.value
+        speed_profile_slow_str = SpeedProfile.SLOW.value
+        self.joint_pos_lim: Dict[Joint, Tuple[float, float]] = {}
+        self.joint_vel_abs_lim: Dict[Joint, Tuple[float, float]] = {}
+
+        # Get the velocity limits for the controllable joints
+        for joint_name in Joint:
```

Comment:
might be nice to break this up into separate functions or something for readability